_delphi_utils_python/tests/test_archive.py

```diff
@@ -74,29 +78,29 @@
         "missing_sample_size": [Nans.NOT_MISSING],
         }),
 
-    # All rows common, but no missing columns
+    # Row 1 same, row 2 deleted, row 3 added, row 4 deleted previously
+    # and present again as nan, row 5 deleted previously and absent from file,
+    # row 6 deleted previously, but new missing has different NA code
```

Comment:
i'm confused; i understood archivediffer expected behavior to be:

1. compute .diff files
2. copy raw unchanged export .csv to cache (new/updated files only)
3. generate new export .csvs from .diff files, including any new annotations marking missing regions as deleted

this would imply that CSVS_BEFORE wouldn't list rows 4, 5, or 6, since it's meant to represent the contents of the cached file, yeah? which wouldn't have the deletion annotations, since the cache grabs its copy of the export file before the deletion annotations are applied.

do i have that wrong?

Reply:
I don't follow step (2). I am supposing that the cache contains a previous value for 21229, the new export doesn't have that row, the difference generates a new deleted row for export, which then becomes the new cached value.

Reply:
[`ArchiveDiffer.run()`](https://github.com/cmu-delphi/covidcast-indicators/blob/fb3d70a578bb118da6756cf47cb64669488aa8f5/_delphi_utils_python/delphi_utils/archive.py#L310-L344) performs four tasks:

1. [`update_cache()`](https://github.com/cmu-delphi/covidcast-indicators/blob/fb3d70a578bb118da6756cf47cb64669488aa8f5/_delphi_utils_python/delphi_utils/archive.py#L387-L406) - make sure the local archivediffer cache is up-to-date with the archival copy. For the S3 differ, this means downloading files from S3.
2. [`diff_exports()`](https://github.com/cmu-delphi/covidcast-indicators/blob/fb3d70a578bb118da6756cf47cb64669488aa8f5/_delphi_utils_python/delphi_utils/archive.py#L201-L258) - compare raw export .csv files with cache .csv files and produce .diff files in the export directory. deletion annotations are added in the .diff file only.
3. [`archive_exports()`](https://github.com/cmu-delphi/covidcast-indicators/blob/fb3d70a578bb118da6756cf47cb64669488aa8f5/_delphi_utils_python/delphi_utils/archive.py#L408-L452) - copy raw export .csv files for new and updated files into cache directory. For the S3 differ, sync any resulting cache directory changes to S3.
4. [`filter_exports()`](https://github.com/cmu-delphi/covidcast-indicators/blob/fb3d70a578bb118da6756cf47cb64669488aa8f5/_delphi_utils_python/delphi_utils/archive.py#L279-L308) - convert .diff files to export .csv files

Reply:
I'm not sure where the mismatch is. This matches my mental model, except that I wouldn't say "deletion annotations are added in the .diff file *only*" - the diffs become export files by nothing more than a renaming in `filter_exports`, so deletion annotations propagate to the export files.

Reply:
Trying to be more explicit about the files:

Day 1:
`/cache/d_signal_20200529.csv`
```
21229,1,0.1,10.0,0,0,0
```

`/receiving/d_signal_20200529.csv`
```
```

`/tmp/archiver/d_signal_20200529.diff`
```
21229,NA,NA,NA,4,4,4
```

`/export/d_signal_20200529.csv`
```
21229,NA,NA,NA,4,4,4
```

`/export/d_signal_20200529.csv` copied to S3 cache bucket

Day 2:
`/cache/d_signal_20200529.csv`
```
21229,NA,NA,NA,4,4,4
```

`/receiving/d_signal_20200529.csv`
```
```

`/tmp/archiver/d_signal_20200529.diff`
Without bug fix:
```
21229,NA,NA,NA,4,4,4
```

With bug fix:
`/tmp/archiver/d_signal_20200529.diff` does not exist, no cache update

Reply:
This is supported by what we saw in https://github.com/cmu-delphi/covidcast-indicators/issues/1520 --

* archivediffer cache in S3 has no entries for 21229, no deletion annotations, not even NA
* archive/successful (which comes from the export files) _does_ have entries for 21229, all set to NA, with no deletion annotations

Reply:
Ah, this makes sense. Ok, let me wrap my head around what fix I need to make.